cpp/src/shim/libcurl.cpp

```diff
@@ -116,19 +119,55 @@ CURL* CurlHandle::handle() noexcept { return _handle.get(); }
 
 void CurlHandle::perform()
 {
-  // Perform the curl operation and check for errors.
-  CURLcode err = curl_easy_perform(handle());
-  if (err != CURLE_OK) {
-    std::string msg(_errbuf);  // We can do this because we always initialize `_errbuf` as empty.
-    std::stringstream ss;
-    ss << "curl_easy_perform() error near " << _source_file << ":" << _source_line;
-    if (msg.empty()) {
-      ss << "(" << curl_easy_strerror(err) << ")";
+  long http_code          = 0;
+  int attempt_count       = 1;
+  int base_delay          = 500;   // milliseconds
+  int max_delay           = 4000;  // milliseconds
+  auto http_max_attempts  = kvikio::defaults::http_max_attempts();
+  auto& http_status_codes = kvikio::defaults::http_status_codes();
+
+  while (attempt_count <= http_max_attempts) {
+    auto err = curl_easy_perform(handle());
+    curl_easy_getinfo(handle(), CURLINFO_RESPONSE_CODE, &http_code);
+
+    // Check if we should retry based on HTTP status code
+    if (std::find(http_status_codes.begin(), http_status_codes.end(), http_code) !=
+        http_status_codes.end()) {
+      // Retry only if one of the specified status codes is returned
+      // TODO: Parse the Retry-After header, if it exists.
+      // TODO: configurable maximum wait.
+      if (attempt_count == http_max_attempts) {
+        std::stringstream ss;
+        ss << "kvikio http_max_attempts_reached. attempts=" << http_max_attempts
+           << " reason=" << http_code;
+        throw std::runtime_error(ss.str());
+      } else {
+        // backoff and retry again. With a base value of 500, we retry after
+        // 500ms, 1s, 2s, 4s, ...
+        int backoff_delay = base_delay * (1 << attempt_count);
+        // up to a maximum of `max_delay` seconds.
+        int delay = std::min(max_delay, backoff_delay);
+
+        attempt_count++;
+        std::cout << "Retrying. reason=" << http_code << " after=" << delay
+                  << " attempt=" << attempt_count << " http_max_attempts=" << http_max_attempts
+                  << std::endl;
```

Comment:
I'd use something human-readable here. We should indicate the unit of milliseconds.
```suggestion
        std::cout << "KvikIO: Retrying HTTP request. Got HTTP code " << http_code << " after " << delay
                  << "ms (attempt " << attempt_count << " of " << http_max_attempts
                  << ")." << std::endl;
```